test/integration/NarrativeTesting.js

```diff
@@ -0,0 +1,189 @@
+/* eslint strict: ["error", "global"] */
+/*global browser, $*/
+'use strict';
+
+const {makeURL, clickWhenReady} = require('./wdioUtils');
+
+function mergeObjects(listOfObjects) {
+    const simpleObjectPrototype = Object.getPrototypeOf({});
+
+    function isSimpleObject(obj) {
+        if (typeof obj !== 'object') {
+            return false;
+        }
+        return Object.getPrototypeOf(obj) === simpleObjectPrototype;
+    }
+
+    function merge(targetObj, sourceObj) {
+        // Copy target, so we don't stomp over original objects.
+        // Note that the source object is not copied, since we don't care if
+        // there is object sharing in the result, we just want to ensure that 
+        // we don't overwrite properties of shared objects.
+        targetObj = JSON.parse(JSON.stringify(targetObj));
+
+        Object.keys(sourceObj).forEach(function (key) {
+            if (isSimpleObject(targetObj[key]) && isSimpleObject(sourceObj[key])) {
+                targetObj[key] = merge(targetObj[key], sourceObj[key]);
+            } else {
+                targetObj[key] = sourceObj[key];
+            }
+        });
+        
+        return targetObj;
+    }
+
+    const objectsToMerge = listOfObjects.map((obj, index) => {
+        if (typeof obj === 'undefined') {
+            return {};
+        }
+        if (!isSimpleObject(obj)) {
+            throw new Error(`Can only merge simple objects, object #${index} is a "${typeof obj}"`);
+        } else {
+            return JSON.parse(JSON.stringify(obj));
+        }
+    });
+    let merged = objectsToMerge[0];
+    for (let i = 1; i < objectsToMerge.length; i += 1) {
+        merged = merge(merged, objectsToMerge[i]);
+    }
+    return merged;
+}
+
+class NarrativeTesting {
+    constructor({testData, caseLabel, timeout}) {
+        this.testData = testData;
+        this.caseLabel = caseLabel;
+        this.timeout = timeout;
+        this.caseData = this.getCaseData();
+    }
+
+    getCaseData() {
+        if (this.caseData) {
+            return this.caseData;
+        }
+        const env = browser.config.testParams.ENV;
+
+        // Note, order is least specific to most specific, so that the more
+        // specific can override the least.
+
+        // Top level test cases provide defaults (non-env-specific) per-case
+        // test data.
+        const caseData = this.testData.cases[this.caseLabel];
+
+        // Each env can establish defaults (e.g. narrative id)
+        const envDefaults = this.testData[env].defaults;
```

Comment:
what happens if the env doesn't exist in `this.testData`?

Reply:
It fails, and you would need to create test data for that environment.
Generally, there should be always be test support for a test environment we are going to test against.
There is no support, generally, for creating or validating test data within the integration tests.
I see this type of functionality as a step in that direction.